theseus/labs/lie/lie_tensor.py

```diff
@@ -0,0 +1,146 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+import warnings
+from typing import Any, List, Optional, Tuple, Union
+
+import torch
+
+from theseus.labs.lie.functional.lie_group import UnaryOperatorOpFnType
+from theseus.labs.lie.functional import se3 as _se3_base
+from theseus.labs.lie.functional import so3 as _so3_base
+from .types import ltype as _ltype
+from .types import SE3, SO3
+
+TensorType = Union[torch.Tensor, "LieTensor"]
+_JFnReturnType = Tuple[List[torch.Tensor], TensorType]
+
+
+class LieTensor:
+    def __init__(self, data_tensor: torch.Tensor, ltype: _ltype):
+        self._t = data_tensor
+        self._fn_lib = {
+            SE3: _se3_base,
+            SO3: _so3_base,
+        }[ltype]
+        self._fn_lib.check_group_tensor(data_tensor)
+        self.ltype = ltype
+
+    def __repr__(self) -> str:
+        return f"LieTensor({self._t}, ltype=lie.{self.ltype})"
+
+    def _check_ltype(self, other: "LieTensor", op_name: str):
+        if other.ltype != self.ltype:
+            raise ValueError("f{op_name} requires both tensors to have same ltype.")
+
+    # Returns a new LieTensor that clones the given data and
+    # has the same ltype as self
+    def clone(self) -> "LieTensor":
```

Comment:
Should we also add a to() that internally calls the torch to()?

Reply:
Yes, we can add that in a separate PR, this is getting too big.

Reply:
Not sure if the same `LieTensor` is good. In my view, `LieTensor` usually refers to something like the tangent vector on a Lie group. I personally prefer `LieGroup` to avoid ambiguity. Maybe we can keep `LieTensor` for the tangent vectors on Lie groups, which are usually a pair of `(group, v)`.